lib/src/annotate.rs

```diff
@@ -0,0 +1,397 @@
+// Copyright 2024 The Jujutsu Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! Methods that allow annotation (attribution and blame) for a file in a
+//! repository.
+//!
+//! TODO: Add support for different blame layers with a trait in the future.
+//! Like commit metadata and more.
+
+use std::collections::HashMap;
+use std::io::BufRead;
+
+use thiserror::Error;
+
+use crate::backend::{BackendError, BackendResult, CommitId, FileId, TreeValue};
+use crate::commit::Commit;
+use crate::diff::{Diff, DiffHunk};
+use crate::fileset::FilesetExpression;
+use crate::graph::{GraphEdge, GraphEdgeType};
+use crate::merged_tree::MergedTree;
+use crate::object_id::ObjectId;
+use crate::repo::{ReadonlyRepo, Repo};
+use crate::repo_path::RepoPath;
+use crate::revset::{RevsetEvaluationError, RevsetExpression, RevsetFilterPredicate};
+use crate::store::Store;
+
+/// Various errors that can arise from annotation
+#[derive(Debug, Error)]
+pub enum AnnotateError {
+    /// the requested file path was not found
+    #[error("Unable to locate file: {0}")]
+    FileNotFound(String),
+    /// the file type is incorrect. Usually a directory was given but a regular
+    /// file is required
+    #[error("File {0} must be a regular file, not a directory")]
+    UnsupportedFileType(String),
+    /// the file is in a conflicted state and can therefore not be annotated
+    /// properly
+    #[error("File {0} is conflicted at commit: {1}")]
+    Conflicted(String, String),
+    /// pass-through of uncaught backend errors
+    #[error(transparent)]
+    BackendError(#[from] BackendError),
+}
+
+/// Annotation results for a specific file
+pub struct AnnotateResults {
+    /// An array of annotation results ordered by line.
+    /// For each value in the array, the commit_id is the commit id of the
+    /// originator of the line and the string is the actual line itself (without
+    /// newline terminators). The vector is ordered by appearance in the
+    /// file
+    pub file_annotations: Vec<(CommitId, String)>,
```

Comment:
Line content should be bytes `Vec<u8>` instead of `String`.

Reply:
Yeah, initially since I used BufRead/lines, everything had to be a string. After changing to split instead, I reverted to Vec<u8>